src/librustc_errors/diagnostic.rs

```diff
@@ -81,7 +87,7 @@ impl Diagnostic {
         Diagnostic::new_with_code(level, None, message)
     }
 
-    pub fn new_with_code(level: Level, code: Option<String>, message: &str) -> Self {
+    pub fn new_with_code(level: Level, code: Option<DiagnosticId>, message: &str) -> Self {
```

Comment:
Do we want to guarantee that lints cannot have diagnostic codes? That doesn't seem like an obviously good thing... although I guess that the lint name can serve as a diagnostic code, which is kind of nice.

Reply:
Diagnostic codes are for errors. Everything else is silenceable in some way and thus a lint. Clippy has a lint list like the error code list of rustc. I don't think having two different identifiers for a lint would be useful. At least I can't think of any use